### ISSUE TITLE
Targeting Issues with Target Turned Enemies Off

### ISSUE THREAD

**Issue by tjayharvey** [GregRos/IEMod.pw#63]
Nexus Reports: 
AE Spells supposed to be restricted to friendlies also hit enemies (Ex. Consecrated Ground)
Every attack misses either with Target Turned Enemies off (one case) or Friendly Fire disabled. 

I can't replicate these with any settings of Disable Friendly Fire or Target Turned Enemies. Any other mods that impact targeting that I should be looking at?

**Comment by tjayharvey** [comment]
Hmm. I must have made a mistake in my original test. Retesting has the first one happen to me whenever Target Turned Enemies is off (Disable Friendly Fire setting doesn't seem to matter). The second seems to happen whenever Target Turned Enemies of off and Disable Friendly Fire is on.

**Comment by tjayharvey** [comment]
Found the problem, I think. The following line in FriendlyRightNowAndAlsoWhenConfused seems to be the problem: 

return targetFaction?.IsFriendly(caster) == true || casterFaction?.IsHostile(target) == true;

This looks like a copy/paste from the function above that had only half of it fixed. When TargetTurnedEnemies is on, hostile characters are never getting here (because they aren't originally friendly).

I tried replacing this line with:
return targetFaction?.IsFriendly(caster) == true || casterFaction?.IsFriendly(target) == true;

and it seems to fix both bugs. I will try charming someone and confirming that hostiles still work when Target Turned Enemies is on.

**Comment by tjayharvey** [comment]
Nope, Target Turned Enemies isn't working. If I use Whisper of Treason and then Talon's Reach (foe-only AE), the charmed enemy is not even a target (it doesn't miss, it just isn't included). Retesting without the change suggests it was broken originally, not as a result of the change I made.

**Comment by tjayharvey** [comment]
Okay, it works with Dominated fine, but not charmed. The culprit seems to be the (non-modded) AIController.GetOriginalTeam() function, which returns the correct Team when Dominated but not when Charmed (instead it returns the current team). This appears to be because the Charmed status effect is not caching the original team properly?

**Comment by tjayharvey** [comment]
And even more info: CharacterStats.GetStatusEffectFromAffliction() is the real problem. If an affliction has multiple status effects (as charmed does - it affects both Accuracy and SwapFaction), the function returns the first one only. This is the status effect that GetOriginalTeam() is checking. Unfortunately, with the current build, the Accuracy penalty seems to be the first effect that Charmed applies, which doesn't keep track of the original team. I have confirmed that the original team is available on Charmed from the appropriate StatusEffect.  This just looks like an Obsidian bug that unfortunately disables the mod. I suggest fixing:
1. Add a new function to CharacterStats to get the swapped faction. This is intelligent enough to look 
   for the "SwapFaction" effect that's sourced by either Charmed or Dominated (I wrote such a function for testing, if you want it).
2. Mod GetOriginalTeam() to call the new function, not GetStatusEffectForAffliction().

I'm actually going to separate this out since it's really nothing to do with the reported bug.
